lib/spack/spack/cmd/external.py

```diff
@@ -0,0 +1,204 @@
+        return [x]
+
+
+def _determine_base_dir(prefix):
+    assert os.path.isdir(prefix)
+    if os.path.basename(prefix) == 'bin':
+        return os.path.dirname(prefix)
+
+
+def _get_external_packages(repo, system_path_to_exe=None):
+    if not system_path_to_exe:
+        system_path_to_exe = _get_system_executables()
+
+    exe_pattern_to_pkgs = defaultdict(list)
+    for pkg in repo.all_packages():
+        if hasattr(pkg, 'executables'):
+            for exe in pkg.executables:
+                exe_pattern_to_pkgs[exe].append(pkg)
+
+    pkg_to_found_exes = defaultdict(set)
+    for path, exe in system_path_to_exe.items():
+        for exe_pattern, pkgs in exe_pattern_to_pkgs.items():
+            if re.search(exe_pattern, exe):
+                for pkg in pkgs:
+                    pkg_to_found_exes[pkg].add(path)
+
+    pkg_to_entries = defaultdict(list)
+    resolved_specs = {}  # spec -> exe found for the spec
+
+    # TODO: iterate through this in a predetermined order (e.g. by package
+    # name) to get repeatable results when there are conflicts.
+    for pkg, exes in pkg_to_found_exes.items():
+        for prefix, exes_in_prefix in _group_by_prefix(exes):
+            # TODO: multiple instances of a package can live in the same
+            # prefix, and a package implementation can return multiple specs
+            # for one prefix, but without additional details (e.g. about the
+            # naming scheme which differentiates them), the spec won't be
+            # usable.
+            if hasattr(pkg, 'determine_spec_details'):
+                specs = _convert_to_iterable(
+                    pkg.determine_spec_details(prefix, exes_in_prefix))
+            else:
+                # Note: packages which do not implement
+                # 'determine_spec_details' should use exact-match regular
+                # expressions
+                specs = [spack.spec.Spec(pkg.name)]
+
+            for spec in specs:
+                pkg_prefix = _determine_base_dir(prefix)
+
+                if not pkg_prefix:
+                    tty.debug("{0} does not end with a 'bin/' directory: it"
+                              " cannot be added as a Spack package"
+                              .format(prefix))
+                    continue
+
+                if spec in resolved_specs:
+                    prior_prefix = ', '.join(resolved_specs[spec])
+
+                    tty.debug(
+                        "Executables in {0} and {1} are both associated"
+                        " with the same spec {2}"
+                        .format(prefix, prior_prefix, str(spec)))
+                    continue
+                else:
+                    resolved_specs[spec] = prefix
+
+                pkg_to_entries[pkg.name].append(
+                    ExternalPackageEntry(spec=spec, base_dir=pkg_prefix))
+            else:
+                tty.debug(
+                    'The following executables in {0} were decidedly not'
```

Comment:
"decidedly" seems like overkill here, but I like the way it reads :).

Reply:
@scheibelp I don't think I understand how this debug information is meant to be triggered. It's executed after the:
```python
for spec in specs:
    ...
```
loop and I don't see any `break` statement in the logic to avoid it under certain conditions. If I turn debug information on:
```console
$ spack -d external find cmake 
==> [2020-04-30-22:32:33.309250, 28329] Imported external from built-in commands
==> [2020-04-30-22:32:33.310512, 28329] Imported external from built-in commands
==> [2020-04-30-22:32:33.311191, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/repos.yaml
==> [2020-04-30-22:32:33.313666, 28329] Reading config file /home/culpo/.spack/linux/repos.yaml
==> [2020-04-30-22:32:33.411763, 28329] '/usr/bin/cmake' '--version'
==> [2020-04-30-22:32:33.416848, 28329] The following executables in /usr/bin were decidedly notpart of the package cmake: /usr/bin/ccmake, /usr/bin/cmake
==> [2020-04-30-22:32:33.417374, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/defaults/packages.yaml
==> [2020-04-30-22:32:33.432310, 28329] Reading config file /home/culpo/PycharmProjects/spack/etc/spack/packages.yaml
```
The debug information is triggered even though:
```
$ cat /home/culpo/.spack/linux/packages.yaml
packages:
  cmake:
    paths:
      cmake@3.17.2: /usr
    version: []
    target: []
    compiler: []
    buildable: true
    modules: {}
    providers: {}
```
I think this part should be something like:
```python
if not specs:
   tty.debug('...')
```
before the loop instead of an `else` clause after it.

Reply:
I thought I was being smart but definitely got that wrong. Thanks!